### ISSUE TITLE
inconsistencies between parseStatement() vs parseExpression()

### ISSUE THREAD

**Issue by dev-ns8** [javaparser/javaparser#2562]
Maybe this is a feature rather than a bug, but it makes it difficult to parse code.  

For some reason above my knowledge, if a statement has a semicolon at the end of it, it can be parsed with parseStatement() just fine. However, if it does not end with a semicolon, parseStatement() will throw an error.

On the other hand, if an expression has a semicolon on the end of it, and you use parseExpression(), this will throw an error but it won't if you remove the semicolon.  

This inconsistency seems very odd to me, and is making it tough to implement my use case.  Can someone please enlighten me on this implementation choice please?

**Comment by edefazio** [comment]
While expressions are valid for evaluating in some contexts (like using groovy or jshell
where the `;` is optional), I *think* the goal here is to build a tool that is consistent with the JLS.
https://docs.oracle.com/javase/specs/

FYI there is an abstraction called an `ExpressionStmt` which is just the expression + ";"

but, in general, there is the consistency that any body of code is made of `Statement`s and not a mixture of `Expression`s and `Statement`s which makes traversing over lines of code intuitive and structured. (i.e. we can ask the method to `getStatement(3)` rather than having some other amalgamation of `Statement`s and `Expression`s

In short, and easy alternative (if you keep accidentally adding ";" to expressions, and omitting the ";" from statements is just write a simple Wrapper around the call to your JavaParser to try catch and retry (adding the ';' or removing the ';')...

**Comment by dev-ns8** [comment]
@edefazio Thank you for your answer, `ExpressionStmt` may be of use to me, I'll take a look at the docs on it.  

I'm just getting started with this library, and I'm finding their are a ton of hidden features to uncover.  Do you buy chance know if there is some tool in the lib that may help me with this use case? 

I'm taking input from a user and writing a new java file with it.  To do so I need to classify the lines of user input so I know where to write it. Is this line a class declaration? A method declaration? A plain old expression? A return statement, etc etc.  

Is there some class that will help me classify these different inputs? Right now I'm using regex to parse the lines and using assumptions based on the language spec to classify each one

Thank you for your help.  Really appreciate it.

**Comment by edefazio** [comment]
If you are given a Java source file and a line number:
You'll need to look at `node.getRange()` its optional... but it will give you the line and column range for each node within the AST. Bare in mind, after you change something (i.e. add code) the numbers DO NOT change... they are more telling you that ... when I parsed this thing where was it in the file.
this might require you to walk the AST and inspect each node's range to find the "where" you are looking for.

If you are allowing a person to just free modify a Java file (and make changes anywhere) that's well a much harder problem... But I dont want to go there unless thats what you are trying to do.

**Comment by dev-ns8** [comment]
@edefazio actually yes lol.  That is what I'm trying to do.  Proving to be a difficult problem indeed.  Like I said, currently reading input line by line and pattern matching on different types of "statements" or "expressions" that could possibly exist in a java source file. 

Let me ask you one more question and then I'll get out of your hair, I promise.

In regards to this library, what is the broadest, and highest level abstraction for any given line of a java source file? Is it an Expression? 

Like: given these three lines, what is the object that encapsulates all three in this library:

`Collections.singletonList("singleton").forEach(System.out::println);`
`public static void main(String[] args){}`
`someObj.removeListener(this);`

Could all three of these lines be classified as `Statements` or what is the best way to encapsulate these to then parse using this lib?

**Comment by edefazio** [comment]
```java
//this is a `Statement` (probably a `ExpressionStmt` with a `MethodCallExpr` inside)
Collections.singletonList("singleton").forEach(System.out::println); 

//this is a `MethodDeclaration` which is a `BodyDeclaration`
public static void main(String[] args){}

//this is another `Statement` (probably a `ExpressionStmt` with a `MethodCallExpr` inside)
someObj.removeListener(this);
```
there is no thing to unify these (3), other than they all inherit from `Node`
but there is no way to put these (3) things "together" in something because they dont make sense at the same "level" within the AST.

(accept for maybe a `NodeList<Node>` or just a plain old `List<Node>` since everything extends Node)  

...IMHO I'd keep the `MethodDeclaration` which is a `BodyDeclaration` implementations separate from the `Statement`s, because they are relevant at different levels of the AST (the Method is added and or removed from Nodes that have a Body (`ClassOrInterfaceDeclaration`, `EnumDeclaration`).
the `Statement`s can only be used WITHIN a `BodyDeclaration` that has a `Block` (i.e. INSIDE a `MethodDeclaration`)

Before going too far, I'd check out the book and read :
https://javaparser.org/inspecting-an-ast/
to get a handle on how things are put together...

**Comment by dev-ns8** [comment]
Thanks a lot for your help @edefazio .  You have definitely cleared some things up for me regarding this lib

**Comment by edefazio** [comment]
I say this only in the sense that it might help you tactically, but I not full hearted endorsing this... but...

if you are just trying to get the job done (some tactical thing) verses writing a tool for other users... you might want to consider using the diff tool I wrote... (the bad part is... its all super alpha software so the API can change and it's not "official" yet...)

it's in the 
https://github.com/org-jdraft/jdraft/tree/master/src/main/java/org/jdraft

Heres a really simple usage example:

```java
public void testDiffClassParts(){
        /**** This is just setting up to classes to diff ****/
        class A { int a; int b=100;              void m(){} A(){} }

        @_rename("A") //this will rename the class from B to A
        class B { int a; int b=200; String name; }

        _class _a = _class.of(A.class);
        _class _b = _class.of(B.class);
        /****************************************************/
        
        _diff _d = _diff.of( _a, _b ); //here we diff the classes
        System.out.println( _d ); //we can print out the diff

        _d.patchLeftToRight(); //we can patch the changes from left to right
        assertEquals(_a, _b);
    }
```
it will print out:

```java
(4) diffs
   - constructor[A()]   - method[m()]   ~ field[b].init   + field[name]
```

think of it as the "Left" is version 1, and the "Right" is version 2... the diff function will tell you:
left constructor is removed in right
`   - constructor[A()]`
left method is removed in right
`   - method[m()]`
there is a change between the init on field b in left and right
`   ~ field[b].init`
left is missing a field name that is in right
`   + field[name]`

each of these (4) diffs are stored separately and they have references to the nodes they refer to...
you can individually decide which diffs to "patch" or as in above, just patch the entire changes from one side to the other with: `_d.patchLeftToRight();`

Here's why I'm recommending AGAINST using it:

- it doesnt have any documentation (you can look at unit tests but thats it)
- I'm too busy to support it fully at the moment (but it might be worth a try to 'git 'er done now and revisit later)
- it does require some understanding of JavaParser, and since you are new it might be a little confusing (i.e. you might be better off using JavaParser first)

Here's why you might want to consider it
- Give er a try, and maybe strike gold, if not, "abandon it" and try it the ground up way (abandon early if you encounter too much friction)

I guess I'm looking at it from this perspective:
trying to do what you want (effectively diffing 2 ASTs) is a (IMHO) very time consuming task, having built the diff tool there are alot of things that are going to be messy unless you are familiar with the inner workings of things... So it MIGHT help, or else you got yourself a lot of work to do.
(sorry to sound discouraging)

**Comment by matozoid** [comment]
@edefazio I would definitely recommend *against* having a confused new user look at yet another tool... Not sure why you keep bringing it up if you clearly say it is more or less experimental and badly supported.
